From 16cd15654cc66acc976dd6c4b4c5249561123d37 Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Wed, 25 Mar 2026 04:08:25 +0800
Subject: [PATCH] cachy: CONFIG_CACHY config hooks (curated backport)
Message-ID: <cachy-fork-cachyos-7.2-rc6-2-config-hooks>

Curated backport of the CachyOS CONFIG_CACHY-gated optimizations from
the CachyOS/linux fork (cachyos-7.2-rc6-2). Individual commits:
  c74130de6b69 Kconfig: Add CONFIG_CACHY
  a9e83bef71e8 CACHY: sched/fair: Tweak EEVDF for interactivity
  f70decbe6bab CACHY: Enable background reclaim of hugepages (THP defrag)
  96b1ae8ac47a CACHY: mm: lru-gen: Protect the working set of the last 100 jiffies
  62f028222e10 CACHY: Disable proactive compaction by default
  488cac6e34bb CACHY: Disable watermark boosting by default
  4107a1558e3f CACHY: Disable split lock mitigation by default

Excluded from the CACHY set: 4cdd54c00ce7 (vm_swappiness=100) and the
elevator bfq-default hunk (single-queue only, irrelevant to our NVMe).

Signed-off-by: Eric Naim <dnaim@cachyos.org>
Signed-off-by: Sultan Alsawaf <sultan@kerneltoast.com>
Signed-off-by: Jan Alexander Steffens (heftig) <jan.steffens@gmail.com>
---
diff --git a/arch/x86/kernel/cpu/bus_lock.c b/arch/x86/kernel/cpu/bus_lock.c
index bba28607a59a..776b083ef603 100644
--- a/arch/x86/kernel/cpu/bus_lock.c
+++ b/arch/x86/kernel/cpu/bus_lock.c
@@ -47,7 +47,11 @@ static const struct {
 
 static struct ratelimit_state bld_ratelimit;
 
+#ifdef CONFIG_CACHY
+static unsigned int sysctl_sld_mitigate = 0;
+#else
 static unsigned int sysctl_sld_mitigate = 1;
+#endif /* CONFIG_CACHY */
 static DEFINE_SEMAPHORE(buslock_sem, 1);
 
 #ifdef CONFIG_PROC_SYSCTL
diff --git a/init/Kconfig b/init/Kconfig
index 10f2013b5321..1b4e59ba4df1 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -218,6 +218,10 @@ config THREAD_INFO_IN_TASK
 
 menu "General setup"
 
+config CACHY
+	bool "Kernel tweaks by CachyOS"
+	default y
+
 config BROKEN
 	bool
 	help
diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index d78467ec6ee1..80ecdcc0a263 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -76,10 +76,19 @@ unsigned int sysctl_sched_tunable_scaling = SCHED_TUNABLESCALING_LOG;
  *
  * (default: 0.70 msec * (1 + ilog(ncpus)), units: nanoseconds)
  */
+#ifdef CONFIG_CACHY
+unsigned int sysctl_sched_base_slice			= 400000ULL;
+static unsigned int normalized_sysctl_sched_base_slice	= 400000ULL;
+#else
 unsigned int sysctl_sched_base_slice			= 700000ULL;
 static unsigned int normalized_sysctl_sched_base_slice	= 700000ULL;
+#endif /* CONFIG_CACHY */
 
+#ifdef CONFIG_CACHY
+__read_mostly unsigned int sysctl_sched_migration_cost	= 400000UL;
+#else
 __read_mostly unsigned int sysctl_sched_migration_cost	= 500000UL;
+#endif
 
 static int __init setup_sched_thermal_decay_shift(char *str)
 {
diff --git a/kernel/sched/sched.h b/kernel/sched/sched.h
index 56acf502ba26..5fac34a2f67b 100644
--- a/kernel/sched/sched.h
+++ b/kernel/sched/sched.h
@@ -3100,7 +3100,7 @@ static inline void attach_one_task(struct rq *rq, struct task_struct *p)
 	attach_task(rq, p);
 }
 
-#ifdef CONFIG_PREEMPT_RT
+#if defined(CONFIG_PREEMPT_RT) || defined(CONFIG_CACHY)
 # define SCHED_NR_MIGRATE_BREAK 8
 #else
 # define SCHED_NR_MIGRATE_BREAK 32
diff --git a/mm/compaction.c b/mm/compaction.c
index f08765ade014..3a94cef84a07 100644
--- a/mm/compaction.c
+++ b/mm/compaction.c
@@ -1903,7 +1903,11 @@ static int sysctl_compact_unevictable_allowed __read_mostly = CONFIG_COMPACT_UNE
  * aggressively the kernel should compact memory in the
  * background. It takes values in the range [0, 100].
  */
+#ifdef CONFIG_CACHY
+static unsigned int __read_mostly sysctl_compaction_proactiveness;
+#else
 static unsigned int __read_mostly sysctl_compaction_proactiveness = 20;
+#endif
 static int sysctl_extfrag_threshold = 500;
 static int __read_mostly sysctl_compact_memory;
 
diff --git a/mm/huge_memory.c b/mm/huge_memory.c
index 58cabe6af33d..3735bfe820c7 100644
--- a/mm/huge_memory.c
+++ b/mm/huge_memory.c
@@ -64,7 +64,11 @@ unsigned long transparent_hugepage_flags __read_mostly =
 #ifdef CONFIG_TRANSPARENT_HUGEPAGE_MADVISE
 	(1<<TRANSPARENT_HUGEPAGE_REQ_MADV_FLAG)|
 #endif
+#ifdef CONFIG_CACHY
+	(1<<TRANSPARENT_HUGEPAGE_DEFRAG_KSWAPD_OR_MADV_FLAG)|
+#else
 	(1<<TRANSPARENT_HUGEPAGE_DEFRAG_REQ_MADV_FLAG)|
+#endif
 	(1<<TRANSPARENT_HUGEPAGE_DEFRAG_KHUGEPAGED_FLAG)|
 	(1<<TRANSPARENT_HUGEPAGE_USE_ZERO_PAGE_FLAG);
 
diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index ee902a468c2f..f5290d209c72 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -270,7 +270,11 @@ const char * const migratetype_names[MIGRATE_TYPES] = {
 
 int min_free_kbytes = 1024;
 int user_min_free_kbytes = -1;
+#ifdef CONFIG_CACHY
+static int watermark_boost_factor __read_mostly;
+#else
 static int watermark_boost_factor __read_mostly = 15000;
+#endif
 static int watermark_scale_factor = 10;
 int defrag_mode;
 
diff --git a/mm/vmscan.c b/mm/vmscan.c
index 56708d1d2dfd..8fce20cdc2f8 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -4128,7 +4128,11 @@ static bool lruvec_is_reclaimable(struct lruvec *lruvec, struct scan_control *sc
 }
 
 /* to protect the working set of the last N jiffies */
+#ifdef CONFIG_CACHY
+static unsigned long lru_gen_min_ttl __read_mostly = 100;
+#else
 static unsigned long lru_gen_min_ttl __read_mostly;
+#endif
 
 static void lru_gen_age_node(struct pglist_data *pgdat, struct scan_control *sc)
 {

